![ExpanGe Bringing Genomic Expansion into focus](https://github.com/mharper10114/ExpanGe/blob/master/media/ExapanGe_logo2.png)
ExpanGe is bringing genomic expansions into focus. A bioinformatics tool to analyze gene expansion

# Introduction

With the increasing availability of chromosome level assemblies investigations of structral variation within newly sequenced genomes  is becoming commonplace. Popular approaches like [JCVI](https://github.com/tanghaibao/jcvi) and [COGE](https://genomevolution.org/coge/) detect syntentic regions between a query and reference genome using genes as anchors for whole genome alignment. [MUMmer](http://mummer.sourceforge.net/) is capable of using a varitiey of methods to determine synteny including a gene free prediction strategy that can be applied independent of annotation. It detects syntey through areas of sequence similiarity between reference and query genomes. Its anchors are called mums. Using MUMmer allows the researcher to employ the different search methods included in MUMmer but still produce a consistent and predictable output.

# Method

`ExpanGe` takes as input the output of MUMmer and calculates the change delta delta between reference and query. It correctly handles calculations for delta delta at the borders of inversions and within inversions.  At present `ExanGe` requires 1 to 1 chromsome correspondance.

## Delta Delta Calculation
![delta delta figure](https://github.com/mharper10114/ExpanGe/blob/master/media/deltadelta2.png)

----------------------------------------
# Making Preliminary Files

## MUMmer

## Running Nucmer for Closely Related Organisms

* Start by running `nucmer`. The `--mum` flag is the most important flag since it will reduce redundancy in sequence matches. The `-p` flag and fasta names are flexible and user defined and will not affect `ExpanGe`

```
nucmer --mum --coords -p A_ref_B_query   A.fasta  B.fasta 
```

* Produce output for `ExpanGE`

```
 show-coords -H -T -r -b -L 100  A_ref_B_query.delta  > A_ref_B_query.HTrbL100.coords
```
  *  `-H` : No header **Required**
  *  `-T` : Tab delimited **Required**
  *  `-r` : Sort by Reference **Required**
  *  `-b` : Non-redundant alignments only **Optional**
  *  `-L <int>` : Filter alignments by length, it must be greater than value **Optional**

------------------------------------

## Gene Map 

Is a user generated, tab separated file that `ExpanGe` uses to determine which mums are are between homologous chromosomes.

```
Example: ref_chrom  ref_len  query_chrom query_len
    chr1  2000  ChR1A  3000
    chr2  4500  ChR3A  6003
    ...
```

* `ref_chrom` :: <str> name given by user can be anything but must exactly match the name that appears in mum file
* `ref_len` :: <int> length of sequence/contig/chromosome
* `query_chrom` :: <str> name given by user can be anything but must exactly match the name that appears in mum file
* `query_len` :: <int>  length of sequence/contig/chromsome




# ExpanGe 
ExpanGe is a bioinformatics tool that is used to analyze gene expansion and contraction.
Using a coordinates file and a gene map file, the program will output the distances between mums
for a reference gene sequence and a query gene sequence. It will then compare the delta values
for the reference and query sequences. In addition, the program identifies where inversions along the
sequences exist and adjusts the calculations accordingly.

```
                
USAGE: python3 ExpanGe.py -i <coordinates file> -g <gene map file> -o <output file> -c <cutoff value> -h <help>

                -h (--help)        Displays help information
                -i (--input)       Input the filename for the coordinates file
                -g (--genemap)     Input the filename for the gene map file
                -o (--output)      Input the filename for the output file that is created
                -c (--cutoff)      Sets the cutoff value for outlier calculations (Default is 1000000)
```
### NOTES:  
* No sorting is done in program, calculations will be made based on order of inputted file. The inputted coordinates file should be sorted in ascending order of the reference starting position column.
* Coordinates file is generated by MUMmer script 'show-coords
* See MUMmer section for instructions on how to produce input.


